google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploy.java

```diff
@@ -42,141 +39,91 @@
-  }
-
-  @Override
-  public void run() {
-    File stagingDirectory = stage();
-
-    if (stagingDirectory != null) {
-      deploy(stagingDirectory);
-    }
   }
 
   /**
-   * Stage the deployment artifacts and return the staging directory or null if it failed.
+   * Given a staging directory, deploy the application to Google App Engine.
    */
-  @Nullable
-  private File stage() {
-    File stagingDirectory;
-    try {
-      stagingDirectory = FileUtil.createTempDirectory(
-          "gae-mvm" /* prefix */,
-          null /* suffix */,
-          true  /* deleteOnExit */);
-      consoleLogLn(
-          "Created temporary staging directory: " + stagingDirectory.getAbsolutePath());
-    } catch (IOException ex) {
-      logger.warn(ex);
-      callback.errorOccurred(
-          GctBundle.message("appengine.deployment.error.creating.staging.directory"));
-      return null;
-    }
-
-    try {
-      File stagedArtifactPath =
-          copyFile(stagingDirectory, "target" + artifactType, deploymentArtifactPath);
-      stagedArtifactPath.setReadable(true /* readable */, false /* ownerOnly */);
-
-      File appYamlPath = deploymentConfiguration.isAuto()
-          ? appEngineHelper.defaultAppYaml()
-          : CloudSdkUtil.getFileFromFilePath(deploymentConfiguration.getAppYamlPath());
-
-      File dockerFilePath = deploymentConfiguration.isAuto()
-          ? appEngineHelper.defaultDockerfile(
-          DeploymentArtifactType.typeForPath(deploymentArtifactPath))
-          : CloudSdkUtil.getFileFromFilePath(deploymentConfiguration.getDockerFilePath());
+  public void deploy(@NotNull File stagingDirectory, @NotNull ProcessStartListener startListener) {
+    final StringBuilder rawDeployOutput = new StringBuilder();
 
-      copyFile(stagingDirectory, "app.yaml", appYamlPath);
-      copyFile(stagingDirectory, "Dockerfile", dockerFilePath);
-    } catch (IOException ex) {
-      logger.warn(ex);
-      callback.errorOccurred(GctBundle.message("appengine.deployment.error.during.staging"));
-      return null;
+    DefaultDeployConfiguration configuration = new DefaultDeployConfiguration();
+    configuration.setDeployables(
+        Collections.singletonList(new File(stagingDirectory, "app.yaml")));
+    configuration.setProject(deploymentConfiguration.getCloudProjectName());
+    configuration.setPromote(true);
+    if (!StringUtil.isEmpty(deploymentConfiguration.getVersion())) {
+      configuration.setVersion(deploymentConfiguration.getVersion());
     }
 
-    return stagingDirectory;
-  }
-
-  /**
-   * Perform a deployment from the given staging directory.
-   */
-  private void deploy(@NotNull File stagingDirectory) {
-    final StringBuilder rawDeployOutput = new StringBuilder();
-    CloudSdk sdk;
-
-    ProcessOutputLineListener stdErrListener = new ProcessOutputLineListener() {
+    ProcessOutputLineListener outputListener = new ProcessOutputLineListener() {
```

Comment:
I wouldn't really understand what the difference between outputListener and deployOutputListener is if I didn't know a bit about the library already.